osh/builtin_meta.py

```diff
@@ -39,18 +39,30 @@
 
 
 class Eval(vm._Builtin):
-    def __init__(self, parse_ctx, exec_opts, cmd_ev, tracer, errfmt):
-        # type: (ParseContext, optview.Exec, CommandEvaluator, dev.Tracer, ui.ErrorFormatter) -> None
+
+    def __init__(self, parse_ctx, exec_opts, cmd_ev, expr_ev, tracer, errfmt):
+        # type: (ParseContext, optview.Exec, CommandEvaluator, expr_eval.ExprEvaluator, dev.Tracer, ui.ErrorFormatter) -> None
         self.parse_ctx = parse_ctx
         self.arena = parse_ctx.arena
         self.exec_opts = exec_opts
         self.cmd_ev = cmd_ev
+        self.expr_ev = expr_ev
         self.tracer = tracer
         self.errfmt = errfmt
 
     def Run(self, cmd_val):
         # type: (cmd_value.Argv) -> int
 
+        if cmd_val.typed_args:
+            t = typed_args.ReaderFromArgv(cmd_val.argv[1:], cmd_val.typed_args,
+                                          self.expr_ev)
```

Comment:
Side note, not strictly related to this PR - I kinda think we should change `cmd_val` so it has evaluated arguments

I was thinking about the lazy evaluation semantics for awhile, and then I realized we want something at the call site

I think it will be

```
json write (x)  # eager
json write [x]  # lazy, synonym for (^[x])

```

there's a Zulip thread on it